src/book.py

```diff
@@ -522,35 +508,54 @@ def _read_kraken_ledgers(self, file_path: Path) -> None:
 
                 row = reader.line_num
 
-                # Skip "duplicate entries" for deposits / withdrawals
-                if _type in dup_state.keys():
-                    skip = dup_state[_type] == dup_skip[_type]
-                    dup_state[_type] = (dup_state[_type] + 1) % 2
-                    if skip:
+                # Skip duplicate entries for deposits / withdrawals and
+                # additional deposit / withdrawals lines for
+                # staking / unstaking / staking reward actions
+                if _type in ["deposit", "withdrawal"]:
+                    if refid not in refids:
+                        refids.append(refid)
                         continue
```

Comment:
Does the behavior changed? Before this PR the second deposit and the first withdrawal entry needed to be skipped. Now its always the first which gets skipped. Is this correct?

Could it happen, that there are more than two transactions with the same refid, so that the first gets skipped and the other ones will run through?

Reply:
Yes, this behavior changed, and I'm not sure yet what the best approach is here (I mentioned it in the PR text).

There are always two entries for deposits/withdrawals: One for crediting the transfer to/from the account, and one for writing it to the blockchain (according to the comments in the code). The action visible to others is the first deposit and second withdrawal. However, for the user of the account, the second deposit and first withdrawal may be more relevant (also for taxing), since this is when the user actually has the assets at their disposal. I think this is especially relevant for the FIFO/LIFO principle (example: you trade an asset while a deposit/withdrawal is ongoing).

What would you suggest here? If we want to consider the first deposit/withdrawal, we could create a "hold" list of operations, and only release them once the second deposit/withdrawal occurs.

For the other actions (staking / unstaking / staking rewards) that lead to "fake" deposits/withdrawals in the CSV exports, the deposits/withdrawals can just be ignored in my opinion (refer to the tables in #97).

Regarding your question, I've never seen more than two transactions with the same refid, except for margin trading.

Reply:
I share your opinion. We should use the user-side transactions for tax calculations.

Do the two transactions differ in timestamps? I thought they are exact the same, but with different types. If that's the case, we might want to assert that the coin pair, value and timestamp are the same, before discarding it regardlessly.

Reply:
Yes, the timestamps for deposits/withdrawals differ (couple of minutes). I've implemented a "hold list" for deposit/withdrawal operations and only release them after assertion when the second action occurs. It works well for the logs that I have, but please review if this implementation also makes sense to you. It would be nice to support multiple export files with the same "hold list", but this would require more implementation effort.

Reply:
see conversation https://github.com/provinzio/CoinTaxman/pull/100#discussion_r800065730